tools/astarte_device_fleet_simulator/lib/astarte_device_fleet_simulator/device.ex

```diff
@@ -0,0 +1,255 @@
+        {:next_state, :waiting, new_data}
+      else
+        # do not wait for all devices
+        Logger.info("#{data.device_id} started sending messages", tag: "device_sending_started")
+        [path: path, value: value, qos: qos] = Config.message_opts()
+
+        {:next_state, :running, new_data,
+         publish_action(Config.publication_interval_ms!(), path, value, qos)}
+      end
+    end
+  end
+
+  def waiting(:cast, :begin_publishing, data) do
+    Logger.info("#{data.device_id} started sending messages", tag: "device_sending_started")
+
+    [path: path, value: value, qos: qos] = Config.message_opts()
+
+    {:next_state, :running, data,
+     publish_action(Config.publication_interval_ms!(), path, value, qos)}
+  end
+
+  def running({:timeout, :publish}, {path, value, qos}, data) do
+    {:ok, device_pid} = fetch_device_pid(data.realm, data.device_id)
+    Task.start(fn -> send_datastream(device_pid, data.interface, path, value, qos) end)
+    {:keep_state_and_data, publish_action(data.publication_interval_ms, path, value, qos)}
+  end
+
+  defp do_register(retry_times, data) do
+    with {:ok, %{body: %{"data" => %{"credentials_secret" => credentials_secret}}}} <-
+           Agent.register_device(data.client, data.device_id),
+         {:ok, _} <-
+           Astarte.Device.start_link(
+             Config.device_opts() ++
+               [device_id: data.device_id, credentials_secret: credentials_secret]
+           ) do
+      {:keep_state,
+       %{
+         client: data.client,
+         device_id: data.device_id,
+         credentials_secret: credentials_secret,
+         realm: data.realm,
+         interface_provider: data.interface_provider,
+         path: data.path,
+         skip_waiting: data.skip_waiting
+       }, connecting_action()}
+    else
+      {:error, reason} ->
+        Logger.error(inspect(reason))
+        backoff_time = compute_backoff_time(retry_times)
+
+        new_retry_times =
+          if retry_times < @max_backoff_exponent, do: retry_times + 1, else: retry_times
+
+        backoff_action = register_action(backoff_time, new_retry_times)
+        {:keep_state_and_data, backoff_action}
+    end
+  end
+
+  defp fetch_device_pid(realm, device_id) do
+    case Device.get_pid(realm, device_id) do
+      nil -> {:error, :unregistered_device}
+      pid -> {:ok, pid}
+    end
+  end
+
+  defp send_datastream(device_pid, interface, path, value, qos) do
+    Logger.debug("Sending #{value} on interface #{interface}, path #{path}, qos #{qos}")
+
+    case Astarte.Device.send_datastream(device_pid, interface, path, value, qos: qos) do
+      :ok -> :ok
+      {:error, reason} -> Logger.error(inspect(reason))
+    end
+  end
+
+  defp publish_action(publication_rate, path, value, qos) do
+    [
+      {{:timeout, :publish}, publication_rate, {path, value, qos}}
+    ]
+  end
+
+  defp connecting_action() do
+    [{:next_event, :internal, :connect}]
```

Comment:
Nitpicking: if you want to provide this utils functions to build FSM actions, it makes sense to _not_ wrap them in the list but let the caller do this, since a state transition could use multiple actions (and wrapping them inside the function means you have to build 2 lists and then concatenate them instead of directly building the list from raw actions)